// koreanbots/dist/src/utils/Collection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const discord_js_1 = require("discord.js");
class LifetimeCollection extends discord_js_1.Collection {
    constructor(options) {
        super();
        this.options = options;
        this.max = options === null || options === void 0 ? void 0 : options.max;
        this.maxAge = options === null || options === void 0 ? void 0 : options.maxAge;
        this.timeouts = new Map();
    }
    set(key, value) {
        if (this.max && super.size >= this.max)
            this.delete(super.randomKey());
        super.set(key, value);
        this.timeouts.set(key, setTimeout(() => {
            super.delete(key);
        }, this.maxAge));
        return this;
    }
    sweep(fn, thisArg) {
        // Copied original method to this.delete points to this.delete instead of super.delete
        if (typeof thisArg !== "undefined")
            fn = fn.bind(thisArg);
        const previousSize = this.size;
        for (const [key, val] of this) {
            if (fn(val, key, this))
                this.delete(key);
        }
        return previousSize - this.size;
    }
    delete(key) {
        const timeout = this.timeouts.get(key);
        if (timeout) {
            clearTimeout(timeout);
            this.timeouts.delete(key);
        }
        return super.delete(key);
    }
    clear() {
        const timeouts = Object.values(this.timeouts);
        for (const timeout of timeouts)
            clearTimeout(timeout);
        this.timeouts.clear();
        super.clear();
    }
}
exports.default = LifetimeCollection;
