// koreanbots/dist/src/utils/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserFlags = exports.BotFlags = void 0;
var BotFlags;
(function (BotFlags) {
    BotFlags[BotFlags["NONE"] = 0] = "NONE";
    BotFlags[BotFlags["OFFICIAL"] = 1] = "OFFICIAL";
    BotFlags[BotFlags["KOREANBOTS_VERIFIED"] = 4] = "KOREANBOTS_VERIFIED";
    BotFlags[BotFlags["PARTNER"] = 8] = "PARTNER";
    BotFlags[BotFlags["DISCORD_VERIFIED"] = 16] = "DISCORD_VERIFIED";
    BotFlags[BotFlags["PREMIUM"] = 32] = "PREMIUM";
    BotFlags[BotFlags["HACKATHON_BOT"] = 64] = "HACKATHON_BOT";
})(BotFlags = exports.BotFlags || (exports.BotFlags = {}));
var UserFlags;
(function (UserFlags) {
    UserFlags[UserFlags["NONE"] = 0] = "NONE";
    UserFlags[UserFlags["ADMINISTRATOR"] = 1] = "ADMINISTRATOR";
    UserFlags[UserFlags["BUG_HUNTER"] = 2] = "BUG_HUNTER";
    UserFlags[UserFlags["BOT_REVIEWER"] = 4] = "BOT_REVIEWER";
    UserFlags[UserFlags["PREMIUM_USER"] = 8] = "PREMIUM_USER";
})(UserFlags = exports.UserFlags || (exports.UserFlags = {}));
