// koreanbots/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Errors = exports.Utils = exports.Constants = exports.Widget = exports.Owners = exports.Github = exports.Discord = exports.Bot = exports.User = exports.Mybot = exports.WidgetManager = exports.UserManager = exports.BotManager = exports.KoreanbotsClient = exports.Koreanbots = void 0;
var Koreanbots_1 = require("./client/Koreanbots");
Object.defineProperty(exports, "Koreanbots", { enumerable: true, get: function () { return Koreanbots_1.Koreanbots; } });
var KoreanbotsClient_1 = require("./client/KoreanbotsClient");
Object.defineProperty(exports, "KoreanbotsClient", { enumerable: true, get: function () { return KoreanbotsClient_1.KoreanbotsClient; } });
var BotManager_1 = require("./managers/BotManager");
Object.defineProperty(exports, "BotManager", { enumerable: true, get: function () { return BotManager_1.BotManager; } });
var UserManager_1 = require("./managers/UserManager");
Object.defineProperty(exports, "UserManager", { enumerable: true, get: function () { return UserManager_1.UserManager; } });
var WidgetManager_1 = require("./managers/WidgetManager");
Object.defineProperty(exports, "WidgetManager", { enumerable: true, get: function () { return WidgetManager_1.WidgetManager; } });
var Mybot_1 = require("./managers/Mybot");
Object.defineProperty(exports, "Mybot", { enumerable: true, get: function () { return Mybot_1.Mybot; } });
var User_1 = require("./structures/User");
Object.defineProperty(exports, "User", { enumerable: true, get: function () { return User_1.User; } });
var Bot_1 = require("./structures/Bot");
Object.defineProperty(exports, "Bot", { enumerable: true, get: function () { return Bot_1.Bot; } });
var Discord_1 = require("./structures/Discord");
Object.defineProperty(exports, "Discord", { enumerable: true, get: function () { return Discord_1.Discord; } });
var Github_1 = require("./structures/Github");
Object.defineProperty(exports, "Github", { enumerable: true, get: function () { return Github_1.Github; } });
var Owners_1 = require("./structures/Owners");
Object.defineProperty(exports, "Owners", { enumerable: true, get: function () { return Owners_1.Owners; } });
var Widget_1 = require("./structures/Widget");
Object.defineProperty(exports, "Widget", { enumerable: true, get: function () { return Widget_1.Widget; } });
exports.Constants = __importStar(require("./utils/Constants"));
exports.Utils = __importStar(require("./utils"));
exports.Errors = __importStar(require("./utils/Errors"));
